### PR TITLE
22579 - Add FED Column & Tooltips (continuation table)

### PR DESCRIPTION
**Pull request by ArwenQin**
*Issue #:*
https://github.com/bcgov/entity/issues/22579

*Description of changes:*
- Added Future Effective Date column to table
- Added red/orange tooltips to future effective date
- Added red/orange tooltips to NR number when it has expired /<14 days
- Added action for NR redirect to Namex UI
- Added hover text showing what statuses are selected
 - For date pickers: show the date range truncated and the full text on hover
 - Added close icon to all search fields
 - Fixed the bug: calendar icon can't open the date picker

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-auth license (Apache 2.0).

### CODE REVIEW

**Comment by ArwenQin** [comment]
Updated the Table name, FED column and actions icon styles:
image

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2978-f8g8fzgr.web.app

**Comment by severinbeauvais** [comment]
I think you should make the Status column wider so it fits without horizontal scrolling. You can make some other columns narrower, such as the dates and the NR number.

**Comment by severinbeauvais** [comment]
Ask Janis if the Name Examination should open in a new page (tab).

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 1]
```diff
@@ -153,9 +282,27 @@
...
+                                  v-bind="attrs"
+                                  v-on="on"
+                                >
+                                  Multiple Selected
+                                </span>
+                              </template>
+                              <span>
+                                <!-- when hover over 'Multiple Selected' showing what are selected. -->
+                                {{
+                                  reviewParams[header.value]
```
Comment (seeker25) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:300]: computed?

**Review by seeker25 — APPROVED** [review]
LGTM, need entities review though

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 1 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Updated

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2978-f8g8fzgr.web.app

**Comment by ArwenQin** [comment]
> Temporary Url for review: https://bcregistry-account-dev--pr-2978-f8g8fzgr.web.app

This link is not the current version, could you pleaser comment /gcbrun again? @severinbeauvais  Thanks!

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2978-f8g8fzgr.web.app

**Comment by severinbeauvais** [comment]
> Updated, made the NR number, date and identifier column narrower. Updated the redirect to NameX, will open in a new tab. 

Looks great!

Since there's an 'X' to clear the dates and Status, what do you think of adding an 'X' to clear the NR Number, Identifying Number and Completing Party?

image

**Comment by severinbeauvais** [comment]
PS You need to "pull up" the icons in the header by 1 or 2 px.

image

image

Also, can you check what the dropdown menu is attached to? It obscures its action button so I have to click away from it to close it.

image

**Comment by severinbeauvais** [comment]
Do you think we should be able to open the NR when the application is approved or rejected?

image

image

I'm thinking "yes" but check with Janis and/or Mihai.

**Comment by ArwenQin** [comment]
> Do you think we should be able to open the NR when the application is approved or rejected?

Do you mean "open Name Request" action? 
I saw a Note "that NRs for Approved/Rejected/Abandoned reviews will not have an expiry date or id or phone/email, so do not show the "open name request" action for these ones."

**Comment by ArwenQin** [comment]
> PS You need to "pull up" the icons in the header by 1 or 2 px.
> 
> image
> 
> image
> 
> Also, can you check what the dropdown menu is attached to? It obscures its action button so I have to click away from it to close it.
> 
> image

It's attached to the second row header <th>

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 2]
```diff
@@ -190,13 +343,41 @@
...
+                      <v-list>
+                        <v-list-item @click="goToNameXUI(item.nrNumber)">
+                          <v-list-item-subtitle>
+                            <v-icon style="font-size: 14px">mdi-format-list-bulleted</v-icon>
+                            <span class="pl-2">Open Name Request</span>
+                          </v-list-item-subtitle>
+                        </v-list-item>
+                      </v-list>
+                    </v-menu>
+                  </span>
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:380]: I found a bug: the sub-menu opens multiple times when I click the down arrow to open it. Double-check this code, or maybe it needs a 'key' prop?.

image

**Comment by severinbeauvais** [comment]
Arwen, do you have any test data to show all 4 warning icon scenarios?

If you prepare which reviews and NRs to update, I can try to help with changing them.

**Comment by ArwenQin** [comment]
> > Updated, made the NR number, date and identifier column narrower. Updated the redirect to NameX, will open in a new tab.
> 
> Looks great!
> 
> Since there's an 'X' to clear the dates and Status, what do you think of adding an 'X' to clear the NR Number, Identifying Number and Completing Party?
> 
> image

Agree. I will add the icons and pull up them

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 2 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:380]: Oh, it seems two reviews have the same NR number, and maybe that's why it's happening.

I don't know if duplicate NRs are valid (eg, it could be for a rejected app and then for a valid app), so if possible let's support this scenario for now.

**Comment by ArwenQin** [comment]
> Arwen, do you have any test data to show all 4 warning icon scenarios?
> 
> If you prepare which reviews and NRs to update, I can try to help with changing them.

I mocked the item.effectiveDate properties in the vue. I'm wondering could you please add some properties in the database?

**Comment by severinbeauvais** [comment]
> Do you mean "open Name Request" action? I saw a Note "that NRs for Approved/Rejected/Abandoned reviews will not have an expiry date or id or phone/email, so do not show the "open name request" action for these ones."

Yes, the Open Name Request action.

I'm not entirely sure why I mentioned "expiry date" in that comment. As for the id or phone/email, that's left over from when we wanted to open the NR in Namerequest UI. Now that we open them in Name Examination UI, those fields are not needed.

I think it can't hurt to provide the ability to open the NR in all cases, subject to what Janis/Mihai say :)

**Comment by severinbeauvais** [comment]
> > Also, can you check what the dropdown menu is attached to? It obscures its action button so I have to click away from it to close it.
> It's attached to the second row header

That sounds right.

You know what? The Status dropdown in the MBR works like this, too, so just leave this one alone for now. No changes needed.

**Comment by severinbeauvais** [comment]
> > If you prepare which reviews and NRs to update, I can try to help with changing them.
> 
> I mocked the item.effectiveDate properties in the vue. I'm wondering could you please add some properties in the database?

Sure. We can do it together. Just let me know which reviews and NRs to update.

**Comment by ArwenQin** [comment]
> Name Examination UI

Thank you.
I prepared 2 testing reviews:
image
These are the last 2 items in the table.
Both are future effective, and the second one has a NR

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 3]
```diff
@@ -71,7 +71,7 @@
                     </div>
                   </IconTooltip>
                   <IconTooltip
-                    v-if="daysLeft(item.effectiveDate) && daysLeft(item.effectiveDate) <= 3"
+                    v-if="daysLeft(item.futureEffectiveDate) && daysLeft(item.futureEffectiveDate) <= 3"
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:74]: This line computes daysLeft twice in some cases. Consider having a single computed instead.

**Comment by ArwenQin** [comment]
- Added close icon to all search fields, moved them up 2 px
 - Fixed the bug: calendar icon can't open the date picker
 - Fixed the bug: action sub-menu opens multiple times
 The status close icon is added by the v-select property "clearable", it seems I can't change its position, is it ok?
image

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 4]
```diff
@@ -82,8 +82,8 @@
                       <strong>Alert:</strong><br>
                       <span>
                         The Future Effective Date for this filing is in
-                        {{ daysLeft(item.effectiveDate) }}
-                        {{ daysLeft(item.effectiveDate) === 1 ? 'day' : 'days' }}.
+                        {{ daysLeft(item.futureEffectiveDate) }}
+                        {{ daysLeft(item.futureEffectiveDate) === 1 ? 'day' : 'days' }}.
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:86]: Similar comment to above. A new computed could return "x days".

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2978-f8g8fzgr.web.app

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 4 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Updated, using daysLeftText()

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 3 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Updated

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2978-f8g8fzgr.web.app

**Comment by severinbeauvais** [comment]
image

This looks very good. Ask Janis if we should show hours instead of days if it's less than, say, 1 day.

NRs always expire at 11:59 pm on their last day, so we probably don't need to show hours for them. What does the tooltip say if the NR expires today?

Sample NR info:
image

**Comment by severinbeauvais** [comment]
### Test results

#### With NR expiring in 6 days (2024-08-29):

image

#### With NR expiring today (2024-08-23):

image

image

Do you think this is OK, or should it show "expiring today"? And also, why isn't the icon red?

**Comment by ArwenQin** [comment]
> ### Test results
> #### With NR expiring in 6 days (2024-08-29):
> image
> 
> #### With NR expiring today (2024-08-23):
> image
> 
> image
> 
> Do you think this is OK, or should it show "expiring today"? And also, why isn't the icon red?

Yes, if the NR expires in less than 24 hour, it will say "expires in 1 day". I think it's ok. Should I check with Janis?
I will check with Janis for the FED less than 24 hours.
The icon is red when the NR already expires (Figma below):
image

**Comment by severinbeauvais** [comment]
> Yes, if the NR expires in less than 24 hour, it will say "expires in 1 day". I think it's ok. Should I check with Janis? I will check with Janis for the FED less than 24 hours. The icon is red when the NR already expires

Yes, check with Janis. As a possible solution, how about if the dates have their own tooltip like this?

image

Yes, you are correct about the red NR icon 👍

**Comment by ArwenQin** [comment]
> > Yes, if the NR expires in less than 24 hour, it will say "expires in 1 day". I think it's ok. Should I check with Janis? I will check with Janis for the FED less than 24 hours. The icon is red when the NR already expires
> 
> Yes, check with Janis. As a possible solution, how about if the dates have their own tooltip like this?
> 
> image
> 
> Yes, you are correct about the red NR icon 👍

Sure, checking with Janis and cc you there.

**Comment by ArwenQin** [comment]
Updated:
- All statuses can open NR
image

- Updated the FED tooltip to show the full date and time
image

- Updated the NR date tooltip expiring today
image

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 5]
```diff
@@ -37,11 +48,95 @@
...
-              <!-- Displaying Formatted Date -->
+              <!-- Displaying Formatted Submission Date -->
               <template #[`item.submissionDate`]="{ item }">
                 <div>{{ formatDate(item.submissionDate) }}</div>
               </template>
 
+              <!-- Displaying Formatted Future Effective Date and Tooltips-->
+              <template #[`item.futureEffectiveDate`]="{ item }">
+                <div>
+                  <v-tooltip bottom>
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:59]: Can you use the same tooltip as elsewhere (which appears above the item and has a little pointer)?

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 6]
```diff
@@ -37,11 +48,95 @@
...
+                    :iconStyling="{'font-size': '1.5em', 'margin-left': '4px'}"
+                    :location="{top: true}"
+                  >
+                    <div>
+                      <strong>Alert:</strong><br>
+                      <span> The Future Effective Date for this filing has passed.</span>
+                    </div>
+                  </IconTooltip>
+                  <IconTooltip
+                    v-if="item.futureEffectiveDate && daysLeft(item.futureEffectiveDate) <= 3"
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:85]: Should we always show the icon and tooltip for a FE date?

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2978-f8g8fzgr.web.app

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 7]
```diff
@@ -68,7 +163,12 @@
                       v-for="(header, i) in headers"
                       :key="getIndexedTag('find-header-row', i)"
                       :scope="getIndexedTag('find-header-col', i)"
-                      class="font-weight-bold"
+                      :class="{
+                        'font-weight-bold': true,
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:167]: No need to change this but FYI: since this class is always set, you can use the `class` prop on this element. (Ie, you can use both `class` and `:class`.)

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 8]
```diff
@@ -190,13 +375,41 @@
                   <span class="open-action">
                     <v-btn
                       color="primary"
-                      :class="['open-action-btn', getButtonLabel(item.status).toLowerCase()]"
+                      :class="['open-action-btn', { active: item.nrNumber }]"
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:378]: I don't understand this syntax. What does the right hand side evaluate to? Should be True or False.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 2 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:412]: This is now fixed. Resolving.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 6 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:85]: Per Figma designs, we only show the icon and tooltip when the FE date passed, and when it's in 3 days, I think it makes sense?

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 9]
```diff
@@ -405,8 +657,58 @@ export default defineComponent({
...
+      } else {
+        const diffDays = moment(effectiveDate).diff(moment(), 'days')
+        return diffDays > 0 ? diffDays : 50
+      }
+    }
+    // Method to display FE days left text
+    const feDaysLeftText = (effectiveDate: string): string => {
+      const diffHours = moment(effectiveDate).diff(moment(), 'hours')
+      if (diffHours > 0 && diffHours <= 24) {
+        return '1 day'
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:690]: I think this should show the number of hours left if it's less than 1 day.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 10]
```diff
@@ -405,8 +657,58 @@ export default defineComponent({
...
+      } else {
+        const diffDays = moment(effectiveDate).diff(moment(), 'days')
+        return `${diffDays} days`
+      }
+    }
+    // Method to display NR days left text
+    const nrDaysLeftText = (effectiveDate: string): string => {
+      const diffHours = moment(effectiveDate).diff(moment(), 'hours')
+      if (diffHours > 0 && diffHours <= 24) {
```
Comment (severinbeauvais) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:699]: I think this assumes that the NR will expire at the end of the day. Would it better to check if the NR expiry day is the same as today?

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 1 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Fixed. Resolving.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 8 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:378]: When the item has a nrNumber, We will adjust the styles of the "Open" Button:
image

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 9 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:690]: Will update it

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 10 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:699]: Will update, thanks

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 5 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:59]: I see why you did this: it's to show the icon tooltip above and the date tooltip below. That's OK with me.

If you keep this, I'd like to see the little arrow, ie:

image

Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:59]: Sure, I will update it

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 6 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:85]: OK. Leave it. Resolving.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 8 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:378]: Please change this to the usual syntax:
```
   class="open-action-btn"
   :class="{ 'active': item.nrNumber }"
```

Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Updated

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 9 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Updated:
image

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 10 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Updated

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue [hunk 11]
```diff
@@ -37,11 +48,97 @@
...
+                      <span
+                        v-bind="attrs"
+                        style="border-bottom: 1px dotted"
+                        v-on="on"
+                      >
+                        {{ formatDate(item.futureEffectiveDate) }}
+                      </span>
+                    </template>
+                    <div class="bottom-tooltip">
+                      <span>{{ formatDate(item.futureEffectiveDate) }} at {{ formatTime(item.futureEffectiveDate) }} Pacific time</span>
```
Comment (ArwenQin) [inline auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: I try to show the little arrow on top of the tooltip, but I'm not sure why this class="bottom-tooltip" didn't work
image

(it's the same style used by the IconTooltip component)

Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: Try without `::v-deep`.

Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: We need to use v-deep when the HTML is rendered by a Vuetify component.

Here, it's "static" HTML in the template.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 9 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Nice! Resolving.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 11 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: I tried without ::v-deep, but it didn't work either

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2978-f8g8fzgr.web.app

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 11 above)
Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: It seems I couldn't change the styles of the tooltip

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 8 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Looks good. Resolving.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 10 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue]: Looks good. Resolving.

File: auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue (hunk 11 above)
Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: I see that `:before` isn't in the DOM. Let me see if I can figure it out.

image

Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: Found it. It's because your CSS is nested in the form. Move it outside and the v-deep will work.

PS You'll have to fix the colour of the pointer.

Comment (ArwenQin) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: Thanks a lot! Fixed it and fixed the color:
image

Comment (severinbeauvais) [reply auth-web/src/components/auth/staff/continuation-application/ContinuationApplicationTable.vue:70]: Looks great! 🎉

File: auth-web/package-lock.json [hunk 12]
```diff
@@ -58,7 +58,7 @@
         "@types/vuelidate": "^0.7.13",
         "@typescript-eslint/eslint-plugin": "^6.4.0",
         "@typescript-eslint/parser": "^6.4.0",
-        "@volar-plugins/vetur": "*",
+        "@volar-plugins/vetur": "latest",
```
Comment (severinbeauvais) [inline auth-web/package-lock.json:61]: Please undo this change.

Also, please update the app version again.

Then, go ahead and merge this PR.

Comment (ArwenQin) [reply auth-web/package-lock.json]: Thanks Sev! I just undo this and updated the app version again.
Could you please merge it?
Thank you.

**Review by severinbeauvais — APPROVED** [review]
